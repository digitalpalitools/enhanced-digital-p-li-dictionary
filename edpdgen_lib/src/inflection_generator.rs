use pls_core::inflections::{generate_inflection_table, PlsInflectionsHost};
use rusqlite::{Connection, Row, NO_PARAMS};

struct PlsHost {}

impl<'a> PlsInflectionsHost<'a> for PlsHost {
    fn get_locale(&self) -> &'a str {
        "en"
    }

    fn get_version(&self) -> &'a str {
        env!("CARGO_PKG_VERSION")
    }

    fn get_url(&self) -> &'a str {
        env!("CARGO_PKG_NAME")
    }

    fn transliterate(&self, s: &str) -> Result<String, String> {
        Ok(s.to_string())
    }

    fn exec_sql_query_core(&self, sql: &str) -> Result<String, String> {
        let table = exec_sql_core(&sql).map_err(|x| x.to_string())?;
        serde_json::to_string(&table).map_err(|x| x.to_string())
    }
}

fn get_row_cells(row: &Row) -> Vec<String> {
    let cells: Vec<String> = row
        .column_names()
        .iter()
        .map(|&cn| {
            let cell: String = match row.get(cn) {
                Ok(c) => c,
                Err(e) => e.to_string(),
            };
            cell
        })
        .collect();

    cells
}

fn exec_sql_core(sql: &str) -> rusqlite::Result<Vec<Vec<Vec<String>>>, rusqlite::Error> {
    let conn = Connection::open("./inflections.db")?;
    let mut result: Vec<Vec<Vec<String>>> = Vec::new();
    for s in sql.split(';').filter(|s| !s.trim().is_empty()) {
        let mut stmt = conn.prepare(&s)?;
        let mut rows = stmt.query(NO_PARAMS)?;

        let mut table: Vec<Vec<String>> = Vec::new();
        while let Some(row) = rows.next()? {
            table.push(get_row_cells(row));
        }
        result.push(table)
    }

    Ok(result)
}

pub trait InflectionGenerator {
    fn generate_inflection_table_html(&self, pali1: &str) -> String;
}

pub(crate) struct NullInflectionGenerator {}

impl NullInflectionGenerator {
    pub fn new() -> NullInflectionGenerator {
        NullInflectionGenerator {}
    }
}

impl InflectionGenerator for NullInflectionGenerator {
    fn generate_inflection_table_html(&self, _pali1: &str) -> String {
        "<div><strong>?????</strong></div>".to_string()
    }
}

pub(crate) struct PlsInflectionGenerator<'a> {
    inflection_host: &'a dyn PlsInflectionsHost<'a>,
}

impl<'a> PlsInflectionGenerator<'a> {
    pub fn new() -> PlsInflectionGenerator<'a> {
        PlsInflectionGenerator {
            inflection_host: &PlsHost {},
        }
    }
}

impl<'a> InflectionGenerator for PlsInflectionGenerator<'a> {
    fn generate_inflection_table_html(&self, pali1: &str) -> String {
        generate_inflection_table(pali1, self.inflection_host).unwrap_or_else(|e| {
            format!(
                "<div>Unable to generate inflection tables. Error: <strong>{}</strong></div>",
                e
            )
        })
    }
}
