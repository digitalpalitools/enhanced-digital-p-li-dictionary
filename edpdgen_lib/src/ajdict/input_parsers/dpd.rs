use crate::ajdict::AjDictPaliWord;
use crate::input::dpd::DpdPaliWord;
use crate::input::make_sort_key;
use tera::{Context, Tera};

lazy_static! {
    static ref TEMPLATES: Tera = {
        let mut tera = Tera::default();
        tera.add_raw_templates(vec![(
            "dpd_word_data",
            include_str!("templates/dpd_word_data.html"),
        )])
        .expect("Unexpected failure adding template");
        tera
    };
}

#[derive(Serialize)]
struct WordDataViewModel<'a> {
    word: &'a DpdPaliWord,
}

impl AjDictPaliWord for DpdPaliWord {
    fn id(&self) -> &str {
        &self.pali1
    }

    fn sort_key(&self) -> String {
        make_sort_key(&self.id())
    }

    fn word_data_entry(&self) -> Result<String, String> {
        let vm = WordDataViewModel { word: &self };

        let context = Context::from_serialize(&vm).map_err(|e| e.to_string())?;
        TEMPLATES
            .render("dpd_word_data", &context)
            .map(|x| x.replace("\r", "").replace("\n", ""))
            .map_err(|e| e.to_string())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::ajdict::input_parsers::load_words;
    use crate::resolve_file_in_manifest_dir;
    use crate::tests::TestLogger;
    use std::path::PathBuf;
    use test_case::test_case;

    pub fn get_csv_path() -> PathBuf {
        resolve_file_in_manifest_dir("Pali_English_Dictionary_10_rows-full.csv")
            .expect("must exist!")
    }

    #[test_case(0)]
    #[test_case(1)]
    #[test_case(2)]
    #[test_case(3)]
    #[test_case(4)]
    #[test_case(5)]
    #[test_case(6)]
    #[test_case(7)]
    #[test_case(8)]
    #[test_case(9)]
    #[test_case(10)]
    #[test_case(11)]
    #[test_case(12)]
    fn word_data_tests(rec_number: usize) {
        let l = TestLogger::new();
        let mut recs = load_words::<DpdPaliWord>(&get_csv_path(), &l).expect("unexpected");

        let word_data = recs
            .nth(rec_number)
            .map(|r| r.word_data_entry().expect("unexpected"))
            .expect("unexpected");

        insta::assert_snapshot!(word_data);
    }
}
