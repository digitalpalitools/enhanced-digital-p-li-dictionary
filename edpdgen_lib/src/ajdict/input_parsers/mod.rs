use crate::ajdict::AjDictPaliWord;
use crate::EdpdLogger;
use regex::{Captures, Regex};
use std::path::Path;

pub mod dpd;

lazy_static! {
    static ref PALI1_CRACKER: Regex = Regex::new(r"(.*)( )(\d+)$").expect("Malformed regex string");
}

//
//
//
//
//  DUP code
//
//
//
//
//
//
pub fn load_words<'a, T: 'a + serde::de::DeserializeOwned + AjDictPaliWord>(
    path: &Path,
    logger: &'a dyn EdpdLogger,
) -> Result<impl Iterator<Item = impl AjDictPaliWord> + 'a, String> {
    logger.info(&format!("Loading words from {:?}.", path));

    let file = std::fs::File::open(path).map_err(|e| e.to_string())?;
    let rdr = csv::ReaderBuilder::new()
        .delimiter(b'\t')
        .flexible(true)
        .from_reader(file);

    let words = rdr
        .into_deserialize::<T>()
        .enumerate()
        .filter_map(move |(i, r)| match r {
            Ok(w) => Some(w),
            Err(e) => {
                logger.error(&format!(
                    "Unable to deserialize record #{}. Error: {}.",
                    i, e
                ));
                None
            }
        });

    logger.info(&format!("... done loading words from {:?}.", &path));
    Ok(words)
}

//
//
//
//
//  DUP code
//
//
//
//
//
//
fn make_sort_key(id: &str) -> String {
    let sk = PALI1_CRACKER.replace(id, |caps: &Captures| {
        // NOTE: Best case effort. Not sweating it.
        let n = &caps[3].parse::<i32>().unwrap_or(0);
        format!("{} {:03}", &caps[1], n)
    });

    sk.into_owned()
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::input::dpd::DpdPaliWord;
    use crate::resolve_file_in_manifest_dir;
    use crate::tests::TestLogger;
    use std::path::PathBuf;
    use test_case::test_case;

    pub fn get_csv_path() -> PathBuf {
        resolve_file_in_manifest_dir("Pali_English_Dictionary_10_rows-full.csv")
            .expect("must exist!")
    }

    #[test_case(0, "훮b훮dha"; "0 digits")]
    #[test_case(9, "adhik훮ra 001"; "1 digit")]
    #[test_case(11, "adhik훮ra 010"; "2 digits")]
    fn test_sort_key(rec_number: usize, expected_sk: &str) {
        let l = TestLogger::new();
        let mut recs = load_words::<DpdPaliWord>(&get_csv_path(), &l).expect("failed to load");

        let sk = recs
            .nth(rec_number)
            .map(|r| make_sort_key(r.id()))
            .expect("unexpected");

        assert_eq!(sk, expected_sk)
    }
}
