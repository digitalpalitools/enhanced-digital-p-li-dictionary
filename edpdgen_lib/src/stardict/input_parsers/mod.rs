use crate::inflection_generator::InflectionGenerator;
use crate::{EdpdLogger, InputFormat};
use regex::{Captures, Regex};
use std::path::Path;

pub mod dpd;
pub mod dps;

lazy_static! {
    static ref PALI1_CRACKER: Regex = Regex::new(r"(.*)( )(\d+)$").expect("Malformed regex string");
}

pub trait PaliWord {
    fn id(&self) -> &str;
    fn sort_key(&self) -> String;
    fn group_id(&self) -> String;
    fn toc_id(&self, input_format: &InputFormat) -> String;
    fn toc_entry(&self, input_format: &InputFormat) -> Result<String, String>;
    fn word_data_entry(
        &self,
        input_format: &InputFormat,
        feedback_form_url: &str,
        host_url: &str,
        host_version: &str,
        igen: &dyn InflectionGenerator,
    ) -> Result<String, String>;
}

pub fn load_words<'a, T: 'a + serde::de::DeserializeOwned + PaliWord>(
    path: &Path,
    logger: &'a dyn EdpdLogger,
) -> Result<impl Iterator<Item = impl PaliWord> + 'a, String> {
    logger.info(&format!("Loading words from {:?}.", path));

    let file = std::fs::File::open(path).map_err(|e| e.to_string())?;
    let rdr = csv::ReaderBuilder::new()
        .delimiter(b'\t')
        .flexible(true)
        .from_reader(file);

    let words = rdr
        .into_deserialize::<T>()
        .enumerate()
        .filter_map(move |(i, r)| match r {
            Ok(w) => Some(w),
            Err(e) => {
                logger.error(&format!(
                    "Unable to deserialize record #{}. Error: {}.",
                    i, e
                ));
                None
            }
        });

    logger.info(&format!("... done loading words from {:?}.", &path));
    Ok(words)
}

fn make_sort_key(id: &str) -> String {
    let sk = PALI1_CRACKER.replace(id, |caps: &Captures| {
        // NOTE: Best case effort. Not sweating it.
        let n = &caps[3].parse::<i32>().unwrap_or(0);
        format!("{} {:03}", &caps[1], n)
    });

    sk.into_owned()
}

fn make_group_id(id: &str) -> String {
    let gid = PALI1_CRACKER.replace(id, |caps: &Captures| caps[1].to_string());

    gid.into_owned()
}

fn make_toc_id(id: &str, input_format: &InputFormat) -> String {
    format!("{}_{}", id.replace(" ", "_"), input_format.to_string())
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::resolve_file_in_manifest_dir;
    use crate::stardict::input_parsers::dpd::DpdPaliWord;
    use crate::tests::TestLogger;
    use std::path::PathBuf;
    use test_case::test_case;

    pub fn get_csv_path() -> PathBuf {
        resolve_file_in_manifest_dir("Pali_English_Dictionary_10_rows-full.csv")
            .expect("must exist!")
    }

    #[test_case(0, "ābādha"; "0 digits")]
    #[test_case(9, "adhikāra 001"; "1 digit")]
    #[test_case(11, "adhikāra 010"; "2 digits")]
    fn test_sort_key(rec_number: usize, expected_sk: &str) {
        let l = TestLogger::new();
        let mut recs = load_words::<DpdPaliWord>(&get_csv_path(), &l).expect("failed to load");

        let sk = recs
            .nth(rec_number)
            .map(|r| make_sort_key(r.id()))
            .expect("unexpected");

        assert_eq!(sk, expected_sk)
    }

    #[test_case(4, "abahulīkata"; "0 digits")]
    #[test_case(5, "abala"; "1 digit")]
    #[test_case(11, "adhikāra"; "2 digits")]
    fn test_group_id(rec_number: usize, expected_gid: &str) {
        let l = TestLogger::new();
        let mut recs = load_words::<DpdPaliWord>(&get_csv_path(), &l).expect("failed to load");

        let gid = recs
            .nth(rec_number)
            .map(|r| make_group_id(r.id()))
            .expect("unexpected");

        assert_eq!(gid, expected_gid);
    }

    #[test_case(4, "abahulīkata_dps"; "0 digits")]
    #[test_case(5, "abala_1_dps"; "1 digit")]
    #[test_case(11, "adhikāra_10_dps"; "2 digits")]
    fn test_toc_id(rec_number: usize, expected_toc_id: &str) {
        let l = TestLogger::new();
        let mut recs = load_words::<DpdPaliWord>(&get_csv_path(), &l).expect("unexpected");

        let toc_id = recs
            .nth(rec_number)
            .map(|r| make_toc_id(r.id(), &InputFormat::DevamittaPaliStudy))
            .expect("unexpected");

        assert_eq!(toc_id, expected_toc_id);
    }
}
