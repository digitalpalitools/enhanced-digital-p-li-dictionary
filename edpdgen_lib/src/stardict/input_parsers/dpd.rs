use crate::inflection_generator::InflectionGenerator;
use crate::stardict::input_parsers::{make_group_id, make_sort_key, make_toc_id, PaliWord};
use tera::{Context, Tera};

lazy_static! {
    static ref TEMPLATES: Tera = {
        let mut tera = Tera::default();
        tera.add_raw_templates(vec![(
            "dpd_toc_summary",
            include_str!("templates/dpd_toc_summary.html"),
        )])
        .expect("Unexpected failure adding template");
        tera.add_raw_templates(vec![(
            "dpd_word_data",
            include_str!("templates/dpd_word_data.html"),
        )])
        .expect("Unexpected failure adding template");
        tera.autoescape_on(vec!["html"]);
        tera
    };
}

// NOTE: Keep the order deliberately randomized as we support column reordering.
#[derive(Debug, Serialize, Deserialize)]
pub struct DpdPaliWord {
    #[serde(rename = "Pāli1")]
    pali1: String,
    #[serde(rename = "Pāli2")]
    pali2: String,
    #[serde(rename = "Fin")]
    fin: String,
    #[serde(rename = "POS")]
    pos: String,
    #[serde(rename = "Grammar")]
    grammar: String,
    #[serde(rename = "Derived from")]
    derived_from: String,
    #[serde(rename = "Neg")]
    neg: String,
    #[serde(rename = "Verb")]
    verb: String,
    #[serde(rename = "Trans")]
    trans: String,
    #[serde(rename = "Case")]
    case: String,
    #[serde(rename = "Meaning IN CONTEXT")]
    in_english: String,
    #[serde(rename = "Sanskrit")]
    sanskrit: String,
    #[serde(rename = "Sk Root")]
    sanskrit_root: String,
    #[serde(rename = "Family")]
    family: String,
    #[serde(rename = "Pāli Root")]
    pali_root: String,
    #[serde(rename = "V")]
    v: String,
    #[serde(rename = "Grp")]
    grp: String,
    #[serde(rename = "Sgn")]
    sgn: String,
    #[serde(rename = "Root Meaning")]
    root_meaning: String,
    #[serde(rename = "Base")]
    base: String,
    #[serde(rename = "Construction")]
    construction: String,
    #[serde(rename = "Derivative")]
    derivative: String,
    #[serde(rename = "Suffix")]
    suffix: String,
    #[serde(rename = "Compound")]
    compound: String,
    #[serde(rename = "Compound Construction")]
    compound_construction: String,
    #[serde(rename = "Source1")]
    source1: String,
    #[serde(rename = "Sutta1")]
    sutta1: String,
    #[serde(rename = "Example1")]
    example1: String,
    #[serde(rename = "Source 2")]
    source2: String,
    #[serde(rename = "Sutta2")]
    sutta2: String,
    #[serde(rename = "Example 2")]
    example2: String,
    #[serde(rename = "Antonyms")]
    antonyms: String,
    #[serde(rename = "Synonyms – different word")]
    synonyms: String,
    #[serde(rename = "Variant – same constr or diff reading")]
    variant: String,
    #[serde(rename = "Commentary")]
    commentary: String,
    #[serde(rename = "Literal Meaning")]
    literal_meaning: String,
    #[serde(rename = "Root In Comps")]
    root_in_compound: String,
    #[serde(rename = "Notes")]
    notes: String,
    #[serde(rename = "Stem")]
    stem: String,
    #[serde(rename = "Pattern")]
    pattern: String,
    #[serde(rename = "Buddhadatta")]
    buddhadatta: String,
    #[serde(rename = "Cl")]
    cl: String,
}

#[derive(Serialize)]
struct WordDataViewModel<'a> {
    word: &'a DpdPaliWord,
    toc_id: &'a str,
    short_name: &'a str,
    feedback_form_url: &'a str,
    host_url: &'a str,
    host_version: &'a str,
    inflection_table: &'a str,
}

impl PaliWord for DpdPaliWord {
    fn id(&self) -> &str {
        &self.pali1
    }

    fn sort_key(&self) -> String {
        make_sort_key(&self.id())
    }

    fn group_id(&self) -> String {
        make_group_id(&self.id())
    }

    fn toc_id(&self, short_name: &str) -> String {
        make_toc_id(&self.id(), short_name)
    }

    fn toc_entry(&self, short_name: &str) -> Result<String, String> {
        let mut context = Context::new();
        context.insert("short_name", short_name);
        context.insert("toc_id", &self.toc_id(short_name));
        context.insert("pali1", &self.pali1);
        context.insert("pos", &self.pos);
        context.insert("in_english", &self.in_english);
        context.insert("buddhadatta", &self.buddhadatta);

        TEMPLATES
            .render("dpd_toc_summary", &context)
            .map_err(|e| e.to_string())
    }

    fn word_data_entry(
        &self,
        short_name: &str,
        feedback_form_url: &str,
        host_url: &str,
        host_version: &str,
        igen: &dyn InflectionGenerator,
    ) -> Result<String, String> {
        let vm = WordDataViewModel {
            word: &self,
            toc_id: &self.toc_id(short_name),
            short_name,
            feedback_form_url,
            host_url,
            host_version,
            inflection_table: &igen.generate_inflection_table_html(&self.pali1),
        };

        let context = Context::from_serialize(&vm).map_err(|e| e.to_string())?;
        TEMPLATES
            .render("dpd_word_data", &context)
            .map_err(|e| e.to_string())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::stardict::input_parsers::load_words;
    use crate::stardict::input_parsers::tests::get_csv_path;
    use crate::tests::{TestInflectionGenerator, TestLogger};
    use test_case::test_case;

    #[test_case(0)]
    #[test_case(1)]
    #[test_case(2)]
    #[test_case(3)]
    #[test_case(4)]
    #[test_case(5)]
    #[test_case(6)]
    #[test_case(7)]
    #[test_case(8)]
    #[test_case(9)]
    #[test_case(10)]
    #[test_case(11)]
    #[test_case(12)]
    fn toc_summary_tests(rec_number: usize) {
        let l = TestLogger::new();
        let mut recs = load_words::<DpdPaliWord>(&get_csv_path(), &l).expect("unexpected");

        let toc_summary = recs
            .nth(rec_number)
            .map(|r| r.toc_entry("dpy").expect("unexpected"))
            .expect("unexpected");

        insta::assert_snapshot!(toc_summary);
    }

    #[test_case(0)]
    #[test_case(1)]
    #[test_case(2)]
    #[test_case(3)]
    #[test_case(4)]
    #[test_case(5)]
    #[test_case(6)]
    #[test_case(7)]
    #[test_case(8)]
    #[test_case(9)]
    #[test_case(10)]
    #[test_case(11)]
    #[test_case(12)]
    fn word_data_tests(rec_number: usize) {
        let l = TestLogger::new();
        let mut recs = load_words::<DpdPaliWord>(&get_csv_path(), &l).expect("unexpected");
        let igen = TestInflectionGenerator::new();

        let word_data = recs
            .nth(rec_number)
            .map(|r| {
                r.word_data_entry("dpd", "fb_url", "host url", "host version", &igen)
                    .expect("unexpected")
            })
            .expect("unexpected");

        insta::assert_snapshot!(word_data);
    }
}
