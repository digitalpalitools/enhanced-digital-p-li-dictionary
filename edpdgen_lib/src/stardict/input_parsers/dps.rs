use crate::input::dps::DpsPaliWord;
use crate::input::make_sort_key;
use crate::stardict::input_parsers::{make_group_id, make_toc_id};
use crate::stardict::StarDictPaliWord;
use pls_core_extras::inflection_generator::InflectionGenerator;
use tera::{Context, Tera};

lazy_static! {
    static ref TEMPLATES: Tera = {
        let mut tera = Tera::default();
        tera.add_raw_templates(vec![(
            "dps_toc_summary",
            include_str!("templates/dps_toc_summary.html"),
        )])
        .expect("Unexpected failure adding template");
        tera.add_raw_templates(vec![(
            "dps_word_data",
            include_str!("templates/dps_word_data.html"),
        )])
        .expect("Unexpected failure adding template");
        tera.autoescape_on(vec!["html"]);
        tera
    };
}

#[derive(Serialize)]
struct WordDataViewModel<'a> {
    word: &'a DpsPaliWord,
    toc_id: &'a str,
    dict_short_name: &'a str,
    feedback_form_url: &'a str,
    host_url: &'a str,
    host_version: &'a str,
    inflection_table: &'a str,
}

impl StarDictPaliWord for DpsPaliWord {
    fn id(&self) -> &str {
        &self.pali
    }

    fn sort_key(&self) -> String {
        make_sort_key(self.id())
    }

    fn group_id(&self) -> String {
        make_group_id(self.id())
    }

    fn toc_id(&self, dict_short_name: &str) -> String {
        make_toc_id(self.id(), dict_short_name)
    }

    fn toc_entry(&self, dict_short_name: &str, concise: bool) -> Result<String, String> {
        let mut context = Context::new();
        context.insert("toc_id", &self.toc_id(dict_short_name));
        context.insert("pali", &self.pali);
        context.insert("fin", &self.fin);
        context.insert("pos", &self.pos);
        context.insert("in_english", &self.in_english);
        context.insert("concise", &concise);

        TEMPLATES
            .render("dps_toc_summary", &context)
            .map_err(|e| e.to_string())
    }

    fn word_data_entry(
        &self,
        dict_short_name: &str,
        feedback_form_url: &str,
        host_url: &str,
        host_version: &str,
        igen: &dyn InflectionGenerator,
        concise: bool,
    ) -> Result<String, String> {
        if concise {
            Ok("".to_string())
        } else {
            let vm = WordDataViewModel {
                word: self,
                toc_id: &self.toc_id(dict_short_name),
                dict_short_name,
                feedback_form_url,
                host_url,
                host_version,
                inflection_table: &igen.generate_inflection_table_html(&self.pali),
            };

            let context = Context::from_serialize(&vm).map_err(|e| e.to_string())?;
            TEMPLATES
                .render("dps_word_data", &context)
                .map_err(|e| e.to_string())
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::resolve_file_in_manifest_dir;
    use crate::stardict::input_parsers::load_words;
    use crate::tests::{TestInflectionGenerator, TestLogger};
    use std::path::PathBuf;
    use test_case::test_case;

    pub fn get_csv_path() -> PathBuf {
        resolve_file_in_manifest_dir("dps_sample.csv").expect("must exist!")
    }

    #[test_case(0, false)]
    #[test_case(1, false)]
    #[test_case(2, false)]
    #[test_case(3, false)]
    #[test_case(4, false)]
    #[test_case(0, true)]
    #[test_case(1, true)]
    #[test_case(2, true)]
    #[test_case(3, true)]
    #[test_case(4, true)]

    fn toc_summary_tests(rec_number: usize, concise: bool) {
        let l = TestLogger::new();
        let mut recs = load_words::<DpsPaliWord>(&get_csv_path(), &l).expect("unexpected");

        let toc_summary = recs
            .nth(rec_number)
            .map(|r| r.toc_entry("dps", concise).expect("unexpected"))
            .expect("unexpected");

        insta::assert_snapshot!(toc_summary);
    }

    #[test_case(0, false)]
    #[test_case(1, false)]
    #[test_case(2, false)]
    #[test_case(3, false)]
    #[test_case(4, false)]
    #[test_case(4, true)]
    fn word_data_tests(rec_number: usize, concise: bool) {
        let l = TestLogger::new();
        let mut recs = load_words::<DpsPaliWord>(&get_csv_path(), &l).expect("unexpected");
        let igen = TestInflectionGenerator::new();

        let word_data = recs
            .nth(rec_number)
            .map(|r| {
                r.word_data_entry("dps", "fb_url", "host url", "host version", &igen, concise)
                    .expect("unexpected")
            })
            .expect("unexpected");

        insta::assert_snapshot!(word_data);
    }
}
