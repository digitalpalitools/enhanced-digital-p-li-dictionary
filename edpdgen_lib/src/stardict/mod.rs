use crate::inflection_generator::InflectionGenerator;
use crate::input::dpd::DpdPaliWord;
use crate::input::dps::DpsPaliWord;
use crate::input::input_format::InputFormat;
use crate::{DictionaryBuilder, DictionaryFile, DictionaryInfo, EdpdLogger};
use std::path::Path;

mod input_parsers;
mod output_generators;

pub struct StarDict<'a> {
    dict_info: &'a DictionaryInfo<'a>,
    input_data_path: &'a Path,
    igen: &'a dyn InflectionGenerator,
    logger: &'a dyn EdpdLogger,
}

impl<'a> DictionaryBuilder<'a> for StarDict<'a> {
    fn new(
        dict_info: &'a DictionaryInfo,
        input_data_path: &'a Path,
        igen: &'a dyn InflectionGenerator,
        logger: &'a dyn EdpdLogger,
    ) -> Self {
        StarDict {
            dict_info,
            input_data_path,
            igen,
            logger,
        }
    }

    fn build_files(&self) -> Result<Vec<DictionaryFile>, String> {
        match self.dict_info.input_format {
            InputFormat::Dpd => run_for_ods_type::<DpdPaliWord>(
                self.dict_info,
                self.input_data_path,
                self.igen,
                self.logger,
            ),
            InputFormat::Dps => run_for_ods_type::<DpsPaliWord>(
                self.dict_info,
                self.input_data_path,
                self.igen,
                self.logger,
            ),
        }
    }
}

pub trait StarDictPaliWord {
    fn id(&self) -> &str;
    fn sort_key(&self) -> String;
    fn group_id(&self) -> String;
    fn toc_id(&self, dict_short_name: &str) -> String;
    fn toc_entry(&self, dict_short_name: &str) -> Result<String, String>;
    fn word_data_entry(
        &self,
        dict_short_name: &str,
        feedback_form_url: &str,
        host_url: &str,
        host_version: &str,
        igen: &dyn InflectionGenerator,
    ) -> Result<String, String>;
}

pub fn run_for_ods_type<'a, T: 'a + serde::de::DeserializeOwned + StarDictPaliWord>(
    dict_info: &DictionaryInfo,
    input_data_path: &Path,
    igen: &dyn InflectionGenerator,
    logger: &dyn EdpdLogger,
) -> Result<Vec<DictionaryFile>, String> {
    let words = input_parsers::load_words::<T>(input_data_path, logger)?;
    let sd_files = output_generators::create_dictionary(&dict_info, words, igen, logger)?;

    Ok(sd_files)
}
