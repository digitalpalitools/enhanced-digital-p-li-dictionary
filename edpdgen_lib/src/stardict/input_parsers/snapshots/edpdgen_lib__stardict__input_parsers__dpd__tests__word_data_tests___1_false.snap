---
source: edpdgen_lib/src/stardict/input_parsers/dpd.rs
expression: word_data

---
  <hr />
  <div>
    <h4 id="ābādhato_dpd" class="word-info-heading-dpd">ābādhato</h4>

    <table class="word-info-table-dpd">
      <tbody><tr><td>Pāli</td><td><span>ābādhato</span></td></tr><tr><td>Grammar</td><td><span>ind, adv, abl sg of ābādha</span></td></tr><tr><td>English</td><td><span><strong>as disease; as sickness; as illness; as affliction</strong></span></td></tr><tr><td>Root</td><td><span>√bādh<sup>･</sup>1 a (to press, oppress)</span></td></tr><tr><td>Construction</td><td><span>ā + √bādh + a + to<br/>ābādha + to</span></td></tr><tr><td>Derivative</td><td><span>taddhita (to)</span></td></tr><tr><td>Sanskrit</td><td><span><i>ābādhāt</i></span></td></tr><tr><td>Sanskrit Root</td><td><span><i>√bādh</i>&nbsp;cl. 1&nbsp;(to oppress)</span></td></tr><tr><td>Commentary</td><td><span>(MNa) rog'aṭṭhena <b>ābādhato</b></span></td></tr></tbody>
    </table>
    <br /><span>so yad'eva tattha hoti rūpagataṃ vedanāgataṃ saññāgataṃ saṅkhāragataṃ viññāṇagataṃ, te dhamme aniccato dukkhato rogato gaṇḍato sallato aghato <b>ābādhato</b> parato palokato suññato anattato samanupassati.</span><br /><span class="sutta-source-dpd"><i>AN 9.36 jhānasuttaṃ</i></span><br /><br /><div>
  <a class="dictionary-feedback-link-dpd" href="fb_url?entry.267696921=host url&entry.1433863141=host version" target="_blank">spot a mistake in the dictionary? fix it here!</a>
</div>
<br />

</div>

<div>
<details>
<summary>Inflection Table</summary>
[ITABLE: ābādhato]
</details>
</div>
<br />

<div>
<a class="dictionary-word-back-to-top-link-dpd" href="#" onclick="document.getElementById('top-dpd').parentElement.previousElementSibling.scrollIntoView();return false;">&#x25B2; back to top</a>
</div>

